elixir/apps/domain/lib/domain/tokens.ex

```diff
@@ -0,0 +1,211 @@
+      |> Authorizer.for_subject(subject)
+      |> Repo.fetch()
+    else
+      false -> {:error, :not_found}
+      other -> other
+    end
+  end
+
+  def list_tokens_for(%Auth.Subject{} = subject) do
+    with :ok <- Auth.ensure_has_permissions(subject, Authorizer.manage_own_tokens_permission()) do
+      Token.Query.by_actor_id(subject.actor.id)
+      |> list_tokens(subject, [])
+    end
+  end
+
+  def list_tokens_for(%Actors.Actor{} = actor, %Auth.Subject{} = subject, opts \\ []) do
+    with :ok <- Auth.ensure_has_permissions(subject, Authorizer.manage_tokens_permission()) do
+      Token.Query.by_actor_id(actor.id)
+      |> list_tokens(subject, opts)
+    end
+  end
+
+  defp list_tokens(queryable, subject, opts) do
+    {preload, _opts} = Keyword.pop(opts, :preload, [])
+
+    {:ok, tokens} =
+      queryable
+      |> Authorizer.for_subject(subject)
+      |> Ecto.Query.order_by([tokens: tokens], desc: tokens.inserted_at, desc: tokens.id)
+      |> Repo.list()
+
+    {:ok, Repo.preload(tokens, preload)}
+  end
+
+  def create_token(attrs) do
+    Token.Changeset.create(attrs)
+    |> Repo.insert()
+  end
+
+  def create_token(attrs, %Auth.Subject{} = subject) do
+    Token.Changeset.create(attrs, subject)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Update allows to extend the token expiration, which is useful for situations where we can use
+  IdP API to refresh the ID token and don't want users to go through redirects every hour
+  (hardcoded token duration for Okta and Google Workspace).
+  """
+  def update_token(%Token{} = token, attrs) do
+    Token.Query.by_id(token.id)
+    |> Token.Query.not_expired()
+    |> Repo.fetch_and_update(with: &Token.Changeset.update(&1, attrs))
+  end
+
+  @doc """
+  Token `secret` is used to verify that token can be used only by one source and that it's
+  impossible to impersonate a session by knowing what's inside our database.
+
+  It then additionally signed and encoded using `Plug.Crypto.sign/3` to make sure that
+  you can't hit our database with requests using a random token id and secret.
+  """
+  def encode_fragment!(%Token{secret_fragment: fragment, type: type} = token)
+      when not is_nil(fragment) do
+    body = {token.account_id, token.id, fragment}
+    config = fetch_config!()
+    key_base = Keyword.fetch!(config, :key_base)
+    salt = Keyword.fetch!(config, :salt)
+    "." <> Plug.Crypto.sign(key_base, salt <> to_string(type), body)
+  end
+
+  def use_token(encoded_token, %Auth.Context{} = context) do
+    with {:ok, {account_id, id, nonce, secret}} <- peek_token(encoded_token, context),
+         queryable =
+           Token.Query.by_id(id)
+           |> Token.Query.by_account_id(account_id)
+           |> Token.Query.by_type(context.type)
+           |> Token.Query.not_expired(),
+         {:ok, token} <- Repo.fetch(queryable),
+         true <-
+           Domain.Crypto.equal?(:sha, nonce <> secret <> token.secret_salt, token.secret_hash) do
```

Comment:
Is this comparison using SHA1?  If so, is that a hard requirement?

Reply:
@bmanifold this is a constant-time comparison with a hash to prevent timing attacks.

Reply:
Replaced with SHA3 256

Reply:
👍